pkg/app/piped/imagewatcher/watcher.go

```diff
@@ -63,108 +65,132 @@ func NewWatcher(cfg *config.PipedSpec, gitClient gitClient, logger *zap.Logger)
+
+	ticker := time.NewTicker(pullInterval)
+	defer ticker.Stop()
 	for {
 		select {
 		case <-ctx.Done():
 			return
 		case <-ticker.C:
-			// Inspect all targets in all repos, and compare only images managed by the given provider.
-			for id, repo := range w.gitRepos {
-				cfg, err := w.loadImageWatcherConfig(ctx, id, repo)
-				if err != nil {
-					w.logger.Error("failed to load image watcher config",
-						zap.String("repo-id", id),
-						zap.Error(err),
-					)
-					continue
-				}
-				for _, target := range cfg.Targets {
-					if target.Provider != provider.Name() {
-						continue
-					}
-					if err := w.updateOutdatedImage(ctx, &target, repo, provider); err != nil {
-						w.logger.Error("failed to update image",
-							zap.String("repo-id", id),
-							zap.String("image-provider", provider.Name()),
-							zap.Error(err),
-						)
-						continue
-					}
-				}
+			err := repo.Pull(ctx, repo.GetClonedBranch())
+			if err != nil {
+				w.logger.Error("failed to perform git pull",
+					zap.String("repo-id", repoCfg.RepoID),
+					zap.String("branch", repo.GetClonedBranch()),
+					zap.Error(err),
+				)
+				continue
+			}
+			cfg, ok, err := config.LoadImageWatcher(repo.GetPath(), includedCfgs, excludedCfgs)
+			if err != nil {
+				w.logger.Error("failed to load configuration file for Image Watcher",
+					zap.String("repo-id", repoCfg.RepoID),
+					zap.Error(err),
+				)
+				continue
+			}
+			if !ok {
+				w.logger.Error("configuration file for Image Watcher not found",
+					zap.String("repo-id", repoCfg.RepoID),
+					zap.Error(err),
+				)
+				continue
+			}
+			if err := w.pushOutdatedImages(ctx, repo, cfg.Targets, commitMsg); err != nil {
+				w.logger.Error("failed to push the changes",
+					zap.String("repo-id", repoCfg.RepoID),
+					zap.Error(err),
+				)
 			}
 		}
 	}
 }
 
-// loadImageWatcherConfig gives back an Image Watcher Config for the given repo.
-func (w *watcher) loadImageWatcherConfig(ctx context.Context, repoID string, repo git.Repo) (*config.ImageWatcherSpec, error) {
-	w.mu.Lock()
-	err := repo.Pull(ctx, repo.GetClonedBranch())
-	w.mu.Unlock()
+// pushOutdatedImages inspects all targets and pushes to git repo after commting the changes.
+func (w *watcher) pushOutdatedImages(ctx context.Context, repo git.Repo, targets []config.ImageWatcherTarget, commitMsg string) error {
+	// Copy the repo to another directory to avoid pull failure in the future.
+	tmpDir, err := ioutil.TempDir("", "image-watcher")
 	if err != nil {
-		return nil, fmt.Errorf("failed to perform git pull: %w", err)
+		return fmt.Errorf("failed to create a new temporary directory: %w", err)
 	}
-
-	var includes, excludes []string
-	for _, repos := range w.config.ImageWatcher.Repos {
-		if repos.RepoID == repoID {
-			includes = repos.Includes
-			excludes = repos.Excludes
-			break
-		}
-	}
-	cfg, ok, err := config.LoadImageWatcher(repo.GetPath(), includes, excludes)
+	defer os.RemoveAll(tmpDir)
+	tmpRepo, err := repo.Copy(tmpDir)
```

Comment:
This `Copy` is done every 5 minutes even if there are no images to update.
I think it is a waste.
It would be better to check the outdated images based on the original read-only repo and then if there are any images to update, let copy to do.

Reply:
Right, will fix it.